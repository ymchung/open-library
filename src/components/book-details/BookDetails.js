import React, { useEffect } from 'react';
import { shape, string, number, bool, func } from 'prop-types';
import classNames from 'classnames';
import { useParams } from 'react-router-dom';

import Loading from '../loading';

const namespace = 'book-details';

const Book = ({ book, isLoading, fetchSelectedBook }) => {
    let { bookId } = useParams();

    useEffect(() => {
        fetchSelectedBook(bookId);
    }, [bookId]);

    return (
        <div className={classNames(namespace, 'row')}>
            {isLoading && <Loading />}
            {!isLoading && book && (
                <>
                    <div className={classNames(`${namespace}__image-container`, 'col-md-3')}>
                        {book.imageUrl ? (<img src={book.imageUrl} className={`${namespace}__image`} alt="Portada del libro"/>) :
                        (<div className={`${namespace}__no-image`}>IMAGEN NO DISPONIBLE</div>)}
                    </div>
                    <div className={classNames(`${namespace}__description-container`, 'col-md-9')}>
                        {book.title && <h3 className={`${namespace}__title`}>{book.title}</h3>}
                        {book.authorName && <p className={`${namespace}__author`}>escrito por <i>{book.authorName}</i></p>}
                        {book.firstPublishYear && <p className={`${namespace}__publish-year`}>Publicado por primera vez en {book.firstPublishYear}</p>}
                        {book.description && (
                            <div className={`${namespace}__info-container`}>
                                <h4 className={`${namespace}__info-container-title`}>Descripci√≥n del libro</h4>
                                <p className={`${namespace}__info-container-text`}>{book.description}</p>
                            </div>
                        )}
                        {book.authorBio && (
                            <div className={`${namespace}__info-container`}>
                                <h4 className={`${namespace}__info-container-title`}>Sobre el autor</h4>
                                <p className={`${namespace}__info-container-text`}>{book.authorBio}</p>
                            </div>
                        )}
                    </div>
                </>
            )}
        </div>
    );
};

Book.propTypes = {
    book: shape({
        title: string,
        firstPublishYear: number,
        authorName: string,
        imageUrl: string,
        authorKey: string,
        description: string,
        authorBio: string,
    }),
    isLoading: bool,
    fetchSelectedBook: func.isRequired,
};

Book.defaultProps = {
    book: null,
    isLoading: true,
};

export default Book;

