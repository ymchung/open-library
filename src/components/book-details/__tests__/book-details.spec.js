import { render } from '@testing-library/react';
import BookDetails from '../BookDetails';

describe('BookDetails Component', () => {
    const props = {
        book: {
            title: 'Harry Potter',
            firstPublishYear: 1999,
            authorName: 'J. K. Rowling',
            imageUrl: 'https:domain.org/image-size.png',
            authorKey: 'A123B',
            description: 'Harry Potter #3 For Harry Potter, it’s the start of another far-from-ordinary year at Hogwarts when the Knight Bus crashes through the darkness and comes to an abrupt halt in front of him.',
            authorBio: 'Joanne "Jo" Murray, OBE (née Rowling), better known under the pen name J. K. Rowling, is a British author best known as the creator of the Harry Potter fantasy series',
        },
        isLoading: true,
        fetchSelectedBook: jest.fn(),
    }

    test('Component should render correctly', () => {
        const { container } = render(<BookDetails {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});