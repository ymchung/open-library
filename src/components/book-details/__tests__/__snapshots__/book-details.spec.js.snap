// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails Component Component should render correctly 1`] = `
<div
  class="book-details row"
>
  <div
    class="loading"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="visually-hidden"
      >
        Cargando...
      </span>
    </div>
  </div>
</div>
`;
